core/system/src/page_allocator.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* This file can be compiled externally to provide the page allocator algorithm
+ * for devices NOT supported by uVisor. For this purpose this file is copied as
+ * is into the target build folder and compiled by the target build system. */
+#if defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1)
+
+#include <uvisor.h>
+#include "page_allocator.h"
+#include "mpu/vmpu_unpriv_access.h"
+#include "mpu/vmpu.h"
+#include "halt.h"
+#include "context.h"
+#define page_table_read  vmpu_unpriv_uint32_read
+#define page_table_write vmpu_unpriv_uint32_write
+
+#else /* defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1) */
+
+#include "uvisor-lib/uvisor-lib.h"
+#define DPRINTF(...) {}
+#define g_active_box 0
+#define vmpu_is_box_id_valid(...) 0
+#define vmpu_public_flash_addr(...) 1
+#define vmpu_sram_addr(...) 1
+#define HALT_ERROR(id, ...) DPRINTF(__VA_ARGS__)
+
+static inline int vmpu_is_region_size_valid(uint32_t size)
+{
+    /* Minimal 32B alignment */
+    return ((size & ~31) == size);
+}
+static inline uint32_t vmpu_round_up_region(uint32_t addr, uint32_t size)
+{
+    if (!vmpu_is_region_size_valid(size)) {
+        return 0;
+    }
+    const uint32_t mask = size - 1;
+    return (addr + mask) & ~mask;
+}
+static inline uint32_t page_table_read(uint32_t addr)
+{
+    return *((uint32_t *) addr);
+}
+static inline void page_table_write(uint32_t addr, uint32_t data)
+{
+    *((uint32_t *) addr) = data;
+}
+
+#endif /* defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1) */
+
+/* By default a maximum of 16 pages are allowed. This can only
+ * be decided by the porting engineer for the current platform! */
```

Comment:
Add comments about the rationale or things a porting engineer must consider when decided on this value. Explain the rationale behind choosing 16.